core/lib/zksync_core/src/basic_witness_input_producer/mod.rs

```diff
@@ -0,0 +1,203 @@
+use std::sync::Arc;
+use std::time::Instant;
+
+use zksync_config::configs::chain::{NetworkConfig, StateKeeperConfig};
+use zksync_dal::ConnectionPool;
+use zksync_object_store::{ObjectStore, ObjectStoreFactory};
+use zksync_queued_job_processor::JobProcessor;
+use zksync_types::witness_block_state::WitnessBlockState;
+use zksync_types::{L1BatchNumber, L2ChainId};
+
+use async_trait::async_trait;
+use tokio::task::JoinHandle;
+
+mod vm_interactions;
+
+use vm_interactions::{create_vm, execute_tx, get_miniblock_transition_state};
+
+#[derive(Debug)]
+pub struct BasicWitnessInputProducerJob {
+    l1_batch_number: L1BatchNumber,
+}
+
+pub struct BasicWitnessInputProducer {
+    connection_pool: ConnectionPool,
+    validation_computational_gas_limit: u32,
+    l2_chain_id: L2ChainId,
+    object_store: Arc<dyn ObjectStore>,
+}
+
+impl BasicWitnessInputProducer {
+    pub async fn new(
+        connection_pool: ConnectionPool,
+        store_factory: &ObjectStoreFactory,
+        state_keeper_config: &StateKeeperConfig,
+        network_config: &NetworkConfig,
+    ) -> anyhow::Result<Self> {
+        // BEWARE, HERE BE DRAGONS.
+        // This usage here is a race condition waiting to happen.
+        // Context -- The VM run in BasicWitnessInputProducer MUST be identical to run from StateKeeper.
+        // Race condition scenario:
+        // 1. StateKeeper runs with a config A, which has value a for `validation_computational_gas_limit`
+        // 2. StateKeeper runs the batch and saves data to database.
+        // 3. Deployment is made and changes to config B, with value b for `validation_computational_gas_limit`.
+        // 4. BasicWitnessInputProducer loads config B
+        // 5. BasicWitnessInputProducer runs the L1Batch that was saved with config A using config B.
+        // In this scenario, the outputs will be different, which will make Witness Generation crash.
+        // This has been discussed between @evl and @deniallugo as a terrible solution, but best so far.
+        let validation_computational_gas_limit =
+            state_keeper_config.validation_computational_gas_limit;
```

Comment:
`validation_computational_gas_limit` in state keeper is used to reject some transactions. In this component you only re-execute non-rejected transactions, so I think it makes sense to set this limit to `<TYPE>::MAX`. We do so in EN IIRC. This also resolves race condition